### PR TITLE
Firestore supported matchingengine

### PR DESCRIPTION
**Pull request by Majidbadal**
So far the matching engine vector store module only supported GCS as backend storage. I have added support for firestore as it is a nicer option for document storage. The corresponding matching engine notebook example is adjusted accordingly.

#### Who can review?
Tag maintainers/contributors who might be interested:
@dev2049

### CODE REVIEW

**Comment by Majidbadal** [comment]
@dev2049 I am not sure about the missing checks. However, the code should work fine and can be tested with a matching engine instance already deployed. 
thanks.

**Review by hwchase17 — COMMENTED** [review]
is thsi better to have as a separate implementation?

they could still share a lot of logic, but it seems like kinda one or the other so maybe better as two vectorstore implementations

**Comment by Majidbadal** [comment]
@hwchase17 thanks for the comment. I am not sure about that because the underlying vector store i.e. vertex's matching engine is the same. The only difference is the backend storage for the documents which would be either GCS or Firestore.
What do you think?

File: langchain/vectorstores/matching_engine.py [hunk 1]
```diff
@@ -80,19 +87,24 @@ def __init__(
...
+        self.firestore_collection_name = firestore_collection_name
+        if not (firestore_collection_name or gcs_bucket_name):
+            raise ValueError("Storage not specified. Either a firestore collection or gcs bucket must be provided!")
 
     def _validate_google_libraries_installation(self) -> None:
         """Validates that Google libraries that are needed are installed."""
         try:
             from google.cloud import aiplatform, storage  # noqa: F401
             from google.oauth2 import service_account  # noqa: F401
+            from google.cloud import firestore
```
Comment (baskaryan) [inline langchain/vectorstores/matching_engine.py:103]: if someone was using gas and didn't have firestore installed, this would raise an error right?

Comment (Majidbadal) [reply langchain/vectorstores/matching_engine.py]: @baskaryan Good catch! fixed now.

**Comment by lkuligin** [comment]
Maybe it's a good idea to separate an implementation of a Matching Engine (that stores vectores and corresponding metadata like ids) and additional mixins that should store a mappings from doc_id to the text chunk.

**Comment by leo-gan** [comment]
@Majidbadal Hi , could you, please, resolve the merging issues and address the last comments (if needed)? After that, ping me and I push this PR for the review. Thanks!
If this PR is not needed anymore, could you, please, let me know or close this PR?

**Comment by Majidbadal** [comment]
@leo-gan Hi,
About the last comment, I don't think a separate implementation is necessary as all mixins and metadata can be stored in firestore using the current implementation.
I resolved the last merge conflict. Please point me to any current issues. Thanks!

**Pull request closed without merging**